tests/havannah-state-tests.cc

```diff
@@ -21,19 +21,20 @@ namespace Havannah {
  template <int SIZE, bool PIE, bool EXTENDED> class StateTest :
    public Havannah::State<SIZE, PIE, EXTENDED> {
   public:
-   StateTest<SIZE, PIE, EXTENDED>(int seed, int history, bool turnFeatures) :
-   Havannah::State<SIZE, PIE, EXTENDED>(seed) {
-      // nhowe added
-      // currently, actively, trying to get the turnFeatures to work properly given it's deprecated
-      // std::list<core::FeatureOptions> featopts;
-      // featopts.emplace_back();
-      // core::FeatureOptions& opt = featopts.back();
-      // opt.turnFeaturesMultiChannel = turnFeatures;
-      // state_->setFeatures(&opt);
+   StateTest<SIZE, PIE, EXTENDED>(int seed, int history, bool turnFeatures)
+       : Havannah::State<SIZE, PIE, EXTENDED>(seed)
+      , _testFeatureOptions(std::make_shared<core::FeatureOptions>()) {
```

Comment:
Allocate pointer to a FeatureOptions for this test state, that will be freed later.

Reply:
So the advantage of this is that it will be freed later, as opposed to never being deallocated?

Reply:
- If you allocate it in the stack (by just creating a `core::FeatureOptions opts` and then passing `&opts`) then yes it is deallocated. But it dies when the function that created it returns, which is bad, as the Test object may persist (it doesn't in this case)

- If you allocate it with a `std::shared_ptr<core::FeatureOptions>`, then so long as one copy of the `std::shared_ptr` is still alive, the object will be in memory.

- If you allocate it with `new core::FeatureOptions` then yes, it's never freed unless you `delete` it explicitly

Reply:
Is this just a matter of having "good form" then? I don't see why we would need it after it's been used in the constructor.

Reply:
`_featopts` (which `setFeatures` sets) is used at various points by the state class (see e.g. `fillFullFeatures` in `state.cc`). If you let the object be deallocated, `_featopts` becomes an invalid pointer, which is bad.

USually in games the feature options are held by `  std::list<FeatureOptions> featopts;` in `core/game.h`, which do get deallocated.

Reply:
You should *never* leave invalid pointers in the program state.